plenum/server/view_change/instance_change_provider.py

```diff
@@ -0,0 +1,118 @@
+import time
+from typing import NamedTuple, Callable
+
+from common.serializers.serialization import instance_change_db_serializer
+from plenum.common.messages.node_messages import InstanceChange
+from storage.kv_store import KeyValueStorage
+from stp_core.common.log import getlogger
+
+logger = getlogger()
+
+
+Vote = NamedTuple("Vote", [
+    ("timestamp", float),
+    ("reason", int)])
+
+
+class InstanceChangeCache(dict):  # Dict[viewNo, Dict[nodeName, Vote]]
+
+    def add(self, view_no, voter, vote: Vote):
+        self.setdefault(view_no, {})
+        self[view_no][voter] = vote
+
+    def remove_vote(self, view_no, voter):
+        if view_no not in self or voter not in self[view_no]:
+            return
+        del self[view_no][voter]
+        if not self[view_no]:
+            del self[view_no]
+
+
+class InstanceChangeProvider:
+
+    def __init__(self, outdated_ic_interval: int = 0,
+                 instance_change_db: KeyValueStorage = None,
+                 time_provider: Callable = time.perf_counter):
+        self._outdated_ic_interval = outdated_ic_interval
+        self._cache = InstanceChangeCache()
+        self._time_provider = time_provider
+        self._instance_change_db = instance_change_db
+        self._fill_cache_by_db()
+
+    def add_vote(self, msg: InstanceChange, voter: str):
+        view_no = msg.viewNo
+        vote = Vote(timestamp=self._time_provider(),
+                    reason=msg.reason)
+        # add to cache
+        self._cache.add(view_no, voter, vote)
+        # add to db
+        self._update_db_from_cache(view_no)
+
+    def has_view(self, view_no: int) -> bool:
+        self._update_votes(view_no)
+        return view_no in self._cache
+
+    def has_inst_chng_from(self, view_no: int, voter: str) -> bool:
+        self._update_votes(view_no)
+        return view_no in self._cache and voter in self._cache[view_no]
+
+    def has_quorum(self, view_no: int, quorum: int) -> bool:
+        self._update_votes(view_no)
+        return view_no in self._cache and len(self._cache[view_no]) >= quorum
+
+    def remove_view(self, view_to_remove: int):
+        for view_no in sorted(self._cache.keys()):
+            if view_no > view_to_remove:
+                break
+            del self._cache[view_no]
+            if self._instance_change_db:
+                self._instance_change_db.remove(str(view_no))
+
+    def items(self):
+        return self._cache.items()
+
+    def _update_votes(self, view_no: int):
+        if self._outdated_ic_interval <= 0 or view_no not in self._cache:
+            return
+        db_need_update = False
+        for voter, vote in dict(self._cache[view_no]).items():
```

Comment:
If I understand correctly `dict` is used here to allow mutation (`self._cache.remove_vote`) during iteration. However `dict` also have `copy()` method, which can help show intention a bit better: `self._cache[view_no].copy().items()`. However this is highly subjective, I don't insist on change.

Reply:
I don't agree, but I believe you :)
Changed